File: apps/mobile/ios/Runner/Native/Volume/VolumeHostApiImpl.swift
import AVKit
import UIKit

final class VolumeHostApiImpl: NSObject, VolumeHostApi {
  private let flutterApi: VolumeFlutterApi
  private var interaction: AnyObject?
  private var listening = false

  init(flutterApi: VolumeFlutterApi) {
    self.flutterApi = flutterApi
    super.init()
  }

  func isAvailable(completion: @escaping (Result<Bool, Error>) -> Void) {
    if #available(iOS 17.2, *) {
      completion(.success(true))
    } else {
      completion(.success(false))
    }
  }

  func startListening() throws {
    setListening(true)
  }

  func stopListening() throws {
    setListening(false)
  }

  private func setListening(_ enabled: Bool) {
    listening = enabled
    guard #available(iOS 17.2, *) else { return }
    if enabled { ensureInteraction() }
    (interaction as? AVCaptureEventInteraction)?.isEnabled = enabled
  }

  @available(iOS 17.2, *)
  private func ensureInteraction() {
    guard interaction == nil, let view = hostView() else { return }
    let eventInteraction = AVCaptureEventInteraction(
      primary: { [weak self] event in
        guard event.phase == .ended else { return }
        self?.emit(.down)
      },
      secondary: { [weak self] event in
        guard event.phase == .ended else { return }
        self?.emit(.up)
      },
    )
    eventInteraction.isEnabled = false
    view.addInteraction(eventInteraction)
    interaction = eventInteraction
  }

  private func emit(_ direction: VolumeDirection) {
    guard listening else { return }
    flutterApi.onVolumePressed(direction: direction) { _ in }
  }

  private func hostView() -> UIView? {
    UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .flatMap(\.windows)
      .first(where: \.isKeyWindow)?
      .rootViewController?.view
  }
}

File: apps/mobile/android/app/src/main/kotlin/com/rarocamera/raro_mobile/volume/VolumeHostApiImpl.kt
package com.rarocamera.raro_mobile.volume

import android.view.KeyEvent
import com.rarocamera.raro_mobile.generated.volume.VolumeDirection
import com.rarocamera.raro_mobile.generated.volume.VolumeFlutterApi
import com.rarocamera.raro_mobile.generated.volume.VolumeHostApi

class VolumeHostApiImpl(
  private val flutterApi: VolumeFlutterApi,
) : VolumeHostApi {
  @Volatile
  var isListening: Boolean = false
    private set

  override fun isAvailable(callback: (Result<Boolean>) -> Unit) {
    callback(Result.success(true))
  }

  override fun startListening() {
    isListening = true
  }

  override fun stopListening() {
    isListening = false
  }

  fun onKeyEvent(event: KeyEvent): Boolean {
    if (!isListening) return false
    val direction =
      when (event.keyCode) {
        KeyEvent.KEYCODE_VOLUME_UP -> VolumeDirection.UP
        KeyEvent.KEYCODE_VOLUME_DOWN -> VolumeDirection.DOWN
        else -> return false
      }
    if (event.action == KeyEvent.ACTION_DOWN && event.repeatCount == 0) {
      flutterApi.onVolumePressed(direction) {}
    }
    return true
  }
}
